### PR TITLE
fix(node): use referrer to determine *.d.ts script kind

### LINKED ISSUE #24451: LSP doesn't resolve npm:stable-hash default export type
Version: Deno 1.44.4

see the [example](https://github.com/guy-borderless/lsp-bug2) here

when go-to-definition on the function I see
```typescript
export default function stableHash(arg: any): string;
```

### LINKED ISSUE #25359: Next.js components from `"next"` package are not importing types properly
I was excited to try out deno on Next.JS after hearing about it on the syntax podcast. It runs well but unfortunately I'm getting a couple of type errors when importing the next core components like `next/link` and `next/image`.

This only happens in the editor with the LSP so it might not be the right place to post it so let me know.

For good measure in the project that I found it in I tried copy and pasting the compiler options into `deno.json`. It had no effect.

> JSX element type 'Link' does not have any construct or call signatures.deno-ts(2604)
> 'Link' cannot be used as a JSX component.
  Its type 'typeof import("file:///Users/richardvanbergen/code/deno-nextjs-bug-example/node_modules/.deno/next@14.2.7/node_modules/next/link")' is not a valid JSX element type.deno-ts(2786)

> JSX element type 'Image' does not have any construct or call signatures.deno-ts(2604)
'Image' cannot be used as a JSX component.
>  Its type 'typeof import("file:///Users/richardvanbergen/code/deno-nextjs-bug-example/node_modules/.deno/next@14.2.7/node_modules/next/image")' is not a valid JSX element type.deno-ts(2786)

Reproduction Steps:

1. Download the repo from the link below
2. Enable Deno from the menu command bar `Deno: Enable`
3. Ensure Deno Future is checked in the user settings
4. Just for good measure, add `DENO_FUTURE=1` explicitly in the settings.
5. Navigate to `src/app/page.tsx`.
6. Behold! Errors!

Version: Deno 1.46.2
Reproduction repo: [https://github.com/richardvanbergen/deno-nextjs-bug-example](https://github.com/richardvanbergen/deno-nextjs-bug-example)

Side note: I can deal with it but is there a way to enable wildcard import aliases?
Side side note: There's also an error importing the styles I personally don't care about that but if it's easy to answer then it might be worth documenting it.

### PR DESCRIPTION
**Pull request by nayeemrmn**
Fixes #24451.
Fixes #25359.

### CODE REVIEW

File: ext/node/ops/worker_threads.rs [hunk 1]
```diff
@@ -66,7 +66,7 @@ where
   match node_resolver.url_to_node_resolution(url)? {
     NodeResolution::Esm(u) => Ok(u.to_string()),
     NodeResolution::CommonJs(u) => wrap_cjs(u),
-    NodeResolution::BuiltIn(_) => Err(generic_error("Neither ESM nor CJS")),
+    _ => Err(generic_error("Neither ESM nor CJS")),
```
Comment (dsherret) [inline ext/node/ops/worker_threads.rs:69]: Can you revert this to not use a wildcard pattern match? When we add or change variants on `NodeResolution` we should get a compiler error here.

File: tests/specs/check/cjs_default_export/main.out [hunk 2]
```diff
@@ -2,6 +2,6 @@ Download http://localhost:4260/@denotest/cjs-default-export
 Download http://localhost:4260/@denotest/cjs-default-export/1.0.0.tgz
 Check file:///[WILDCARD]/main.ts
 error: TS2322 [ERROR]: Type 'number' is not assignable to type 'string'.
-export const Test: string = cjsDefault.default();
+export const Test: string = cjsDefault();
```
Comment (dsherret) [inline tests/specs/check/cjs_default_export/main.out:5]: This should be `cjsDefault.default()` for ESM importing CJS.

**Review by dsherret — CHANGES_REQUESTED** [review]
The fix here is not exactly correct. I think the problem here is that some packages will have package.json exports that lead to a .d.ts file that is in a position that makes it cjs, but it should be treated as esm? I think we need a way to figure out when that happens for .d.ts files only, which will make our node resolution more complicated. I think it requires a refactor of ext/node_resolution/resolver.rs to essentially do `url_to_node_resolution` at a lower level.

**Comment by nayeemrmn** [comment]
I tried the various reproductions which spurred this patch again under Node/TS, this time remembering to add `{ "type": "module" }` in `package.json` and using the following Deno-like tsconfig:
```json
{
  "compilerOptions": {
    "allowJs": true,
    "esModuleInterop": true,
    "experimentalDecorators": false,
    "isolatedModules": true,
    "lib": ["esnext"],
    "module": "nodenext",
    "moduleDetection": "force",
    "noEmit": true,
    "resolveJsonModule": true,
    "strict": true,
    "target": "esnext",
    "useDefineForClassFields": true,
    "jsx": "react",
    "jsxFactory": "React.createElement",
    "jsxFragmentFactory": "React.Fragment",
    "types": ["@types/node"],
  },
}
```
It turns out that none of them are behaving wrong. All of them behave the same as tsc for both the diagnostics shown and the emit/execution. I'll follow up in the issues. Closing.

**Pull request closed without merging**